### PR TITLE
Content artifcats upgrade

### PR DESCRIPTION
**Pull request by teizenman**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/31660
related: https://github.com/demisto/etc/issues/31465

## Must have
- [X] Tests
- [ ] Documentation

### CODE REVIEW

File: demisto_sdk/commands/common/constants.py [hunk 1]
```diff
@@ -1006,3 +1001,97 @@ class PB_Status:
...
+    PLAYBOOKS_DIR,
+    INDICATOR_FIELDS_DIR,
+    REPORTS_DIR,
+    INDICATOR_TYPES_DIR,
+    LAYOUTS_DIR,
+    CLASSIFIERS_DIR,
+    WIDGETS_DIR
+}
+
+CORE_PACKS_LIST = [
```
Comment (yaakovi) [inline demisto_sdk/commands/common/constants.py:1070]: no need, download the file from the content repo (as we do in validate):
```python
tools.get_remote_file('Tests/Marketplace/core_packs_list.json') or []
```

File: demisto_sdk/commands/common/update_id_set.py [hunk 2]
```diff
@@ -748,6 +749,19 @@ def get_report_data(path):
     return parse_dashboard_or_report_data(layouts, report_data, path)
 
 
+def get_metadata_data(path):
```
Comment (yaakovi) [inline demisto_sdk/commands/common/update_id_set.py:752]: add pack ID too

File: demisto_sdk/commands/common/update_id_set.py [hunk 3]
```diff
@@ -1018,6 +1051,16 @@ def get_general_paths(path, pack_to_create):
     return files
 
 
+def get_pack_metadata_path(pack_to_create):
+    if pack_to_create:
+        path = [pack_to_create, '*', PACKS_PACK_META_FILE_NAME]
```
Comment (yaakovi) [inline demisto_sdk/commands/common/update_id_set.py:1056]:
```suggestion
        path = [pack_to_create, PACKS_PACK_META_FILE_NAME]
```

File: demisto_sdk/commands/common/update_id_set.py [hunk 4]
```diff
@@ -1369,6 +1422,7 @@ def re_create_id_set(id_set_path: Optional[str] = DEFAULT_ID_SET_PATH, pack_to_c
     new_ids_dict['Reports'] = sort(reports_list)
     new_ids_dict['Widgets'] = sort(widgets_list)
     new_ids_dict['Mappers'] = sort(mappers_list)
+    new_ids_dict['Packs'] = {key: packs_dict[key] for key in sorted(packs_dict.keys())}
```
Comment (yaakovi) [inline demisto_sdk/commands/common/update_id_set.py:1425]: ?

File: demisto_sdk/tests/integration_tests/content_create_artifacts_integration_test.py [hunk 5]
```diff
@@ -62,3 +67,57 @@ def test_duplicate_file_failure(mock_git):
...
+    )
+
+    with ChangeCWD(repo.path):
+        with temp_dir() as temp:
+            runner = CliRunner(mix_stderr=False)
+            result = runner.invoke(main, [ARTIFACTS_CMD, '-a', temp, '-p', 'all'])
+
+    assert result.exit_code == 0
+    assert 'Pack1' in result.stderr
+    assert 'Pack2' in result.stderr
```
Comment (yaakovi) [inline demisto_sdk/tests/integration_tests/content_create_artifacts_integration_test.py:123]: if you expect a successful run (error code is 0), why do you expect anything in the stderr?

File: demisto_sdk/tests/integration_tests/content_create_artifacts_integration_test.py [hunk 6]
```diff
@@ -62,3 +67,57 @@ def test_duplicate_file_failure(mock_git):
...
+    )
+
+    with ChangeCWD(repo.path):
+        with temp_dir() as temp:
+            runner = CliRunner(mix_stderr=False)
+            result = runner.invoke(main, [ARTIFACTS_CMD, '-a', temp, '-p', 'Pack1'])
+
+    assert result.exit_code == 0
+    assert 'Pack1' in result.stderr
+    assert 'Pack2' not in result.stderr
```
Comment (yaakovi) [inline demisto_sdk/tests/integration_tests/content_create_artifacts_integration_test.py:96]: if you expect a successful run (error code is 0), why do you expect anything in the stderr?

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 7]
```diff
@@ -35,14 +35,17 @@ def __exit__(self, type_, value, traceback):
         self.fd = None
 
 
-def parse_for_pack_metadata(dependency_graph: nx.DiGraph, graph_root: str) -> tuple:
+def parse_for_pack_metadata(dependency_graph: nx.DiGraph, graph_root: str, complete_data: bool = False,
+                            id_set_data: dict = {}) -> tuple:
```
Comment (yaakovi) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:39]: don't use dict as a default value in functions.
```suggestion
                            id_set_data: dict = None) -> tuple:
```

File: demisto_sdk/commands/find_dependencies/find_dependencies.py [hunk 8]
```diff
@@ -1126,7 +1141,10 @@ def find_dependencies(pack_name: str, id_set_path: str = '', exclude_ignored_dep
             dependency_graph = PackDependencies.build_dependency_graph(
                 pack_id=pack_name, id_set=id_set, verbose_file=verbose_file,
                 exclude_ignored_dependencies=exclude_ignored_dependencies)
-        first_level_dependencies, _ = parse_for_pack_metadata(dependency_graph, pack_name)
+        if not complete_data:
```
Comment (yaakovi) [inline demisto_sdk/commands/find_dependencies/find_dependencies.py:1144]: why do you need this condition? the function already handles it

**Review by yaakovi — COMMENTED** [review]
phase 1: haven't reviewed the main changes yet

File: demisto_sdk/commands/common/constants.py [hunk 9]
```diff
@@ -862,6 +856,7 @@ class PB_Status:
 XSOAR_SUPPORT_URL = "https://www.paloaltonetworks.com/cortex"
 MARKETPLACE_LIVE_DISCUSSIONS = \
     'https://live.paloaltonetworks.com/t5/cortex-xsoar-discussions/bd-p/Cortex_XSOAR_Discussions'
+MARKETPLACE_MIN_VERSION = '6.0.0'
```
[the remark this answers is not in the document]
Comment (teizenman) [reply demisto_sdk/commands/common/constants.py:859]: We have it in `format_constants.py`, not in the global constants.

File: demisto_sdk/commands/common/update_id_set.py (hunk 4 above)
Comment (teizenman) [reply demisto_sdk/commands/common/update_id_set.py:1425]: Sorting packs by name

File: demisto_sdk/commands/find_dependencies/find_dependencies.py (hunk 7 above)
Comment (teizenman) [reply demisto_sdk/commands/find_dependencies/find_dependencies.py:39]: This is the way to do that for the `get` later.

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 10]
```diff
@@ -48,55 +54,231 @@
...
-        self.cpus = cpus
-        self.execution_start = time.time()
+        self.content_uploadable_zips_path = self.artifacts_path / 'uploadable_packs'
+
+        # inits
         self.content = Content.from_cwd()
+        self.execution_start = time.time()
         self.exit_code = EX_SUCCESS
 
+    def create_content_artifacts(self) -> int:
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:105]: why did you move it inside the object?

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 11]
```diff
@@ -329,7 +511,8 @@ def dump_packs(artifact_manager: ArtifactsManager, pool: ProcessPool) -> List[Pr
     """
     futures = []
     for pack_name, pack in artifact_manager.content.packs.items():
-        if pack_name not in IGNORED_PACKS:
+        if (pack_name in artifact_manager.pack_names or 'all' in artifact_manager.pack_names) and \
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:514]: instead of iterating the whole content and search for the desired pack, maybe switch to iterate over the given pack names. if `all` is specified - keep the current logic.

check how fast it runs for 2 packs, if not significant - you can leave it as is.

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 12]
```diff
@@ -397,9 +601,18 @@ def dump_pack(artifact_manager: ArtifactsManager, pack: Pack) -> ArtifactsReport
...
-        pack.pack_metadata.dump(artifact_manager.content_packs_path / pack.id)
+    if pack.metadata:
+        pack_report += ObjectReport(pack.metadata, content_packs=True)
+        pack.metadata.content_items = content_items_handler.content_items
+        pack.metadata.server_min_version = content_items_handler.server_min_version
+        if artifact_manager.id_set_path:
+            # Dependencies can only be done when id_set file is given.
+            pack.metadata.dependencies = handle_dependencies(pack, artifact_manager.id_set_path)
+        else:
+            logger.info('Skipping dependencies extraction since no id_set file was provided.')
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:612]: switch to warning

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 13]
```diff
@@ -653,6 +867,9 @@ def ArtifactsDirsHandler(artifact_manager: ArtifactsManager):
         delete_dirs(artifact_manager)
         artifact_manager.exit_code = EX_FAIL
     else:
+        sign_packs(artifact_manager)
+        zip_packs(artifact_manager)
+        encrypt_packs(artifact_manager)
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:872]: if `zip_artifacts` is false, those shouldn't run

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 14]
```diff
@@ -705,3 +936,202 @@ def report_artifacts_paths(artifact_manager: ArtifactsManager):
...
                              artifact_manager.content_all_path]:
             logger.info(template.format(artifact_dir))
+
+
+###############################
+# Metadata handling functions #
+###############################
+
+
+def load_user_metadata(pack: Pack) -> Optional[PackMetaData]:
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:946]: should be a static method of `PackMetaData`

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 15]
```diff
@@ -48,55 +54,231 @@
...
+            DASHBOARDS_DIR: self.add_dashboard_as_content_item,
+            INDICATOR_FIELDS_DIR: self.add_indicator_field_as_content_item,
+            INDICATOR_TYPES_DIR: self.add_indicator_type_as_content_item,
+            REPORTS_DIR: self.add_report_as_content_item,
+            LAYOUTS_DIR: self.add_layout_as_content_item,
+            CLASSIFIERS_DIR: self.add_classifier_as_content_item,
+            WIDGETS_DIR: self.add_widget_as_content_item
+        }
+
+    def handle_content_item(self, content_object: ContentObject):
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:162]: add docstring

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 16]
```diff
@@ -48,55 +54,231 @@
...
+        global logger
+
+        content_object_directory = content_object.path.parts[-3]
+        if content_object_directory not in self.content_folder_name_to_func.keys():
+            content_object_directory = content_object.path.parts[-2]
+
+        # if content_object.path.suffix not in CUSTOM_CONTENT_FILE_ENDINGS:
+        #     return
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:170]: remove

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 17]
```diff
@@ -48,55 +54,231 @@
...
+        # reputation in old format aren't supported in 6.0.0 server version
+        if content_object_directory == INDICATOR_TYPES_DIR and not re.match(content_object.path.name,
+                                                                            'reputation-.*.json'):
+            return
+
+        # skip content items that are not displayed in contentItems
+        if content_object_directory not in CONTENT_ITEMS_DISPLAY_FOLDERS:
+            return
+
+        logging.debug(
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:184]: please attach a screenshot of this log

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 18]
```diff
@@ -48,55 +54,231 @@
...
+            f"Iterating over {content_object.path} file and collecting items of {content_object.path.parts[-4]} pack")
+
+        self.server_min_version = max(self.server_min_version, content_object.from_version)
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:187]: if the pack metadata sets a minimum value, don't override it

**Review by yaakovi — CHANGES_REQUESTED** [review]
nice work

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 19]
```diff
@@ -705,3 +936,199 @@ def report_artifacts_paths(artifact_manager: ArtifactsManager):
...
+        artifact_manager (ArtifactsManager): Artifacts manager object.
+        pack (Pack): The pack to sign.
+        zip_pack_path (Path): Path to the pack not encrypted script.
+        encryption_key (str): The encryption key for the packs.
+
+    """
+    current_working_dir = os.getcwd()
+
+    try:
+        os.chdir(artifact_manager.content_uploadable_zips_path)
```
[the remark this answers is not in the document]
Comment (teizenman) [reply demisto_sdk/commands/create_artifacts/content_artifacts_creator.py]: This was the original logic, I've changed to full paths so it's redundant.

**Review by yaakovi — APPROVED** [review]
🥇 
nice work

File: demisto_sdk/__main__.py [hunk 20]
```diff
@@ -330,9 +330,22 @@ def validate(config, **kwargs):
 @click.option('--cpus',
               help='Number of cpus/vcpus availble - only required when os not reflect number of cpus (CircleCI'
                    'allways show 32, but medium has 3.', hidden=True, default=os.cpu_count())
-def create_arifacts(**kwargs) -> int:
+@click.option('-idp', '--id_set_path', help='The full path of id_set.json')
+@click.option('-p', '--pack_names',
+              help=("Packs to create artifacts for. Optional values are: `all` or "
+                    "csv list of packs. "
+                    "Default is set to `all`"),
+              default="all")
```
Comment (glicht) [inline demisto_sdk/__main__.py:338]: Don't think this is good. Think also that a partner may want to use this to create a pack that is upload-able to his local server. I would rather this be explicit.

Also please add support for the `-i` option where a directory can be specified as we have in other comamnds.

File: demisto_sdk/__main__.py [hunk 21]
```diff
@@ -330,9 +330,22 @@ def validate(config, **kwargs):
 @click.option('--cpus',
               help='Number of cpus/vcpus availble - only required when os not reflect number of cpus (CircleCI'
                    'allways show 32, but medium has 3.', hidden=True, default=os.cpu_count())
-def create_arifacts(**kwargs) -> int:
+@click.option('-idp', '--id_set_path', help='The full path of id_set.json')
```
Comment (glicht) [inline demisto_sdk/__main__.py:333]: Why not hidden too? Is this mandatory?

File: demisto_sdk/commands/create_artifacts/content_artifacts_creator.py [hunk 22]
```diff
@@ -696,12 +930,78 @@ def report_artifacts_paths(artifact_manager: ArtifactsManager):
...
+            else:
+                for pack_name in artifact_manager.pack_names:
+                    if pack_name in artifact_manager.content.packs:
+                        pack = artifact_manager.content.packs[pack_name]
+                        dumped_pack_dir = os.path.join(artifact_manager.content_packs_path, pack.id)
+                        futures.append(pool.schedule(pack.sign_pack, args=(logger, dumped_pack_dir,
+                                                                           artifact_manager.signDirectory,
+                                                                           )))
+        wait_futures_complete(futures, artifact_manager)
+        os.remove('keyfile')
```
Comment (glicht) [inline demisto_sdk/commands/create_artifacts/content_artifacts_creator.py:972]: should probably be in a finally clause. To make sure this is deleted even if there is an exception.
